c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+use crate::context::{AnalysisCtxt, Assignment, LTy};
+use crate::labeled_ty::{LabeledTy, LabeledTyCtxt};
+use crate::rewrite::Rewrite;
+use crate::type_desc::{self, Ownership, Quantity};
+use rustc_ast::ast;
+use rustc_hir as hir;
+use rustc_hir::def::{DefKind, Namespace, Res};
+use rustc_hir::def_id::LocalDefId;
+use rustc_hir::intravisit;
+use rustc_hir::Mutability;
+use rustc_middle::hir::nested_filter;
+use rustc_middle::mir::{self, Body, LocalDecl};
+use rustc_middle::ty::print::{FmtPrinter, Print};
+use rustc_middle::ty::subst::GenericArg;
+use rustc_middle::ty::{self, ReErased, Ty, TyCtxt, TyKind};
+use rustc_span::Span;
+
+/// A label for use with `LabeledTy` to indicate what rewrites to apply at each position in a type.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+struct RewriteLabel {
+    /// Rewrite a raw pointer, whose ownership and quantity have been inferred as indicated.
+    ty_desc: Option<(Ownership, Quantity)>,
+    /// If set, a child or other descendant of this type requires rewriting.
+    descendant_has_rewrite: bool,
+}
+
+type RwLTy<'tcx> = LabeledTy<'tcx, RewriteLabel>;
+
+/// Given an `LTy`, which is labeled with `PointerId`s, determine which rewrites to apply based on
+/// the permissions and flags inferred for each `PointerId`.
+fn relabel_rewrites<'tcx>(
+    asn: &Assignment,
+    lcx: LabeledTyCtxt<'tcx, RewriteLabel>,
+    lty: LTy<'tcx>,
+) -> RwLTy<'tcx> {
+    lcx.relabel_with_args(lty, &mut |lty, args| {
+        let ty_desc = if lty.label.is_none() {
+            None
+        } else {
+            let perms = asn.perms()[lty.label];
+            let flags = asn.flags()[lty.label];
+            // TODO: if the `Ownership` and `Quantity` exactly match `lty.ty`, then `ty_desc` can
+            // be `None` (no rewriting required).  This might let us avoid inlining a type alias
+            // for some pointers where no actual improvement was possible.
+            Some(type_desc::perms_to_desc(perms, flags))
+        };
+        let descendant_has_rewrite = args.iter().any(|child| {
+            let has_rewrite = child.label.ty_desc.is_some();
```

Comment:
nit: why not `child.label.ty_desc.is_some() || child.label.descendant_has_rewrite`?

Reply:
Just clarifying that what we're checking here, abstractly, is "the child has a rewrite or one of its descendants has a rewrite".  The expression `child.label.ty_desc.is_some()` is not obviously connected to "child has a rewrite" IMO.

Reply:
I agree; it's clearer this way.